text/0000-const-bounds-methods.md

```diff
@@ -0,0 +1,579 @@
+    fn default() -> Self;
+}
+```
+
+And and some type - for instance:
+
+```rust
+struct Foo(usize);
+```
+
+As a rustacean, you may now write:
+
+```rust
+impl Default for Foo {
+    const fn default() -> Self {
+        Foo(0)
+    }
+}
+```
+
+Note that this `impl` has constrained `default` more than required by the
+`Default` trait. Why this is useful will be made clear in the [motivation]
+and in the [guide-level-explanation] of `const` trait bounds.
+
+Naturally, `default` for `Foo` will now be type-checked as a `const fn`.
+
+## 3. syntactic sugar `const impl`
+
+In any inherent `impl` (not an impl of a trait), or an `impl` of a trait,
+you may now write:
+
+```rust
+const impl MyType {
+    fn foo() -> usize;
+
+    fn bar(x: usize, y: usize) -> usize;
+
+    // ..
+}
+
+const impl MyTrait for MyType {
+    fn baz() -> usize;
+
+    fn quux(x: usize, y: usize) -> usize;
+
+    // ..
+}
+```
+
+and have the compiler desugar this for you into:
+
+```rust
+impl MyType {
+    const fn foo() -> usize;
+
+    const fn bar(x: usize, y: usize) -> usize;
+
+    // ..
+}
+
+impl MyTrait for MyType {
+    const fn baz() -> usize;
+
+    const fn quux(x: usize, y: usize) -> usize;
+
+    // ..
+}
+```
+
+For the latter case of `const impl MyTrait for MyType`, this always means that
+`MyType` may be used to substitute for `T` in a bound like `T: const MyTrait`.
+
+The compiler will of course now check that the `fn`s are const, and refuse to
+compile your program if you lied to the compiler.
+
+When it comes to migrating existing code to this new model, it is recommended
+that you simply start by adding `const` right before your `impl` and see if it
+still compiles and then continue this process until all `impl`s that can be
+`const impl` are. For those that can't, you can still add `const fn` to some
+`fn`s in the `impl`. The standard library will certainely follow this process
+in trying to make the standard library as (re)useable as possible.
+
+## 4. `const` trait bounds, `T: const Trait`
+
+Speaking of const trait bounds, what are they? They are simply a bound-modifier
+on a bound `T: Trait` denoted as `T: const Trait` with some changed semantics.
+
+What are the semantics? That any type you substitute for `T` must in addition
+to impl the trait in question, also do so without any normal `fn`s. Any `fn`s
+occuring in the `impl` must be marked as `const fn`. These `impl`s are exactly
+those `impl`s that are currently, or would type check as `const impl`.
+
+A `const Trait` bound gives you the power to use all `fn`s in the trait in a
+`const` context such as in const generics, `const` bindings and in `const fn`s
+in general.
+
+Currently, this RFC also proposes that you be allowed to write `impl const Trait`
+and `impl const TraitA + const TraitB`, both for static existential and universal
+quantification (return and argument positiion). However, the RFC does not, in
+its current form, mandate the addition of syntax like `Box<const Trait>`.
```

Comment:
I think that `const Trait` objects should be elaborated in a bit more detail. What does it mean to have `&const Trait` or `Box<const Trait>`? Right now, AFAIK, const eval doesn't cover references, so something like:

```
const NUMBER: &usize = 5;
const NUMBER_2: usize = *NUMBER;
```

would fail. If this behaviour were to be changed (and IMO, it makes sense to do so), then something along the lines of `&const Trait` would be reasonably well-formed. That said, `Box<const Trait>` still makes no sense.

Longer term, it might make sense to simply use inference to determine if a function can be executed in constant time or not. For example, take the following (not very efficient) function:

```
const fn pow<T: const Mul + const One>(val: T, exp: usize) {
    if exp == 0 {
        T::one()
    } else {
        val * pow(val, exp - 1)
    }
}
```

To me, it makes more sense to completely forgo the `const Mul + const One` in favour of `Mul + One`. This would allow the function to be called for inputs which are not const which do not support `const fn`, but also let the function be called in a const context if, for example, `T` was `u64`.

Perhaps a clippy lint could warn for calling `const fn`s like this one with a constant value where the requisite traits were not `const` as well. However, in general, it would be allowed by the language.

I feel like this sort of thing would make a lot more sense than simply providing a `const Trait` definition which opens up a whole interesting world of things.

That said, adding `const impl Trait` syntax would be totally fine IMHO. The only issue would be that things like `impl const Default + Iterator` would no longer be expressable, but I'm not sure what the benefit of that would be. At that rate, why not simply define a type directly which has all of your requirements?

Reply:
Also adding further to this: `&const Trait` only makes sense if it's `&'static const Trait`, because `const`s all have `'static` lifetimes now. Which kind of further makes things a bit confusing.

Reply:
Another addendum: `const impl Trait` in function signatures makes sense considering how `const impl Trait { ... }` is allowed by the third option. Which, IMHO, is another good justification for option number three.

Reply:
Thank you for a great and elaborate review!

Good point on `const impl Trait` and the `const impl Trait { .. }` syntax - I hadn't thought of it. I will surely add this to the motivation. Thanks <3.

Writing `Box<const Trait>`, entails that it can only be constructed by some `T` for which `T: const Trait` holds. Other than that, it is just `Box<Trait>`. The same applies to `&(const Trait)`. The only expressive power you have gained with `Box<const Trait>` is the ability to restrict to `const fn`s in the trait impl, nothing else. Since this is marginally useful, I have opted to not propose it yet. Perhaps if not added in this RFC, it can be added in a subsequent RFC. I have written a bit on `&'static const Trait` now.

Type inference is nice, but also more complex (algorithmically) and is considerably more costly wrt. compile times. The fact that Rust already has long compile times is something we should try hard not to increase too much. In addition, there is certain expressive power to the ability of restriction - do note that `T: const Trait` is also allowed for normal `fn`s and `impl<T: const Trait>`. I think the explicitness and uniformity is a good thing.

The `impl const Default + Iterator` thing is fully expressible with this proposal - as an example of what you can use this for is to plug into the universal quantification here:
```rust
fn foo<T: const Default + Iterator>(bar: T) {
    // Use the const Default bound:
    const BAZ: T = T::default();

    // Use the iterator bound:
    BAZ.foreach(|elt| { dbg!(elt); });
}
```

Reply:
As I mentioned, I'm having trouble understanding in what context you'd want a `impl T1 + const T2`; the case you mentioned is worthwhile but I'm not sure if it bears the weight of the syntax implications it holds. As stated, you can always use newtypes to get this kind of expressiveness, but in most cases when you're returning a value you either want it to work in all const contexts or none, IMHO.

Reply:
I'm for sure open to `const impl TraitA + Trait B` as a syntax if `impl const TraitA + const TraitB` is too weird to swallow - but my base position is always that expressivity is more important than syntax, even if syntax may become clunky as a result of it. If there is popular demand for this change I will certainly do it.

Reply:
This may be a supremely bad idea, but you could also do:  `-> const impl TraitA + impl TraitB`.